File: electron/tests/net-invite_test.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// THE INVITE LAUNCH ARGUMENTS — `+connect_lobby <id>` and `--connect <addr>`.
//
// Worth a test despite being twenty lines, because the input is a command line
// somebody ELSE writes: Steam's, when a friend accepts an invite while the game
// is closed. There is no way to try that path from a keyboard, and the failure
// mode is silent — the game opens its title menu as if nothing had happened,
// which is precisely what it did before this existed.

import { describe, expect, it } from "vitest";

import { readInvite } from "../src/net-invite";

/** What a real launch looks like: the binary, then whatever Electron and Steam
 * put after it. The flag is never argv[0], and never the last word either — a
 * parser that only handled the tidy case would pass here and fail in the wild. */
const LAUNCH = ["/opt/game/game", "--no-sandbox"];

describe("reading an invite off a command line", () => {
  it("takes Steam's own +connect_lobby", () => {
    expect(readInvite([...LAUNCH, "+connect_lobby", "109775241010"])).toEqual({
      lobbyId: "109775241010",
    });
  });

  it("takes a shared address", () => {
    expect(readInvite([...LAUNCH, "--connect", "203.0.113.7:27016"])).toEqual({
      address: "203.0.113.7:27016",
    });
  });

  it("ignores a flag with nothing after it", () => {
    // A truncated command line must not produce a join attempt against "" —
    // which would spend six seconds probing an address nobody typed and then
    // report that nobody was hosting there.
    expect(readInvite([...LAUNCH, "+connect_lobby"])).toBeNull();
    expect(readInvite([...LAUNCH, "--connect", "--fullscreen"])).toBeNull();
  });

  it("ignores an ordinary launch", () => {
    expect(readInvite(LAUNCH)).toBeNull();
  });
});

File: tauri/shell/tests/net_invite_test.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
//! `+connect_lobby` AND `--connect` — the two ways a session reaches a game
//! that was not running.

use adastrail_shell::net_invite::{read_invite, Invite};

fn argv(args: &[&str]) -> Vec<String> {
    args.iter().map(|arg| (*arg).to_string()).collect()
}

#[test]
fn steams_own_invite_is_read_off_the_command_line() {
    // A friend accepted while the game was closed; Steam launches the binary
    // with this and nothing else says a join was asked for.
    assert_eq!(
        read_invite(&argv(&["+connect_lobby", "109775240983270000"])),
        Some(Invite::Lobby("109775240983270000".to_string()))
    );
}

#[test]
fn a_shared_address_is_read_in_either_spelling() {
    for flag in ["--connect", "+connect"] {
        assert_eq!(
            read_invite(&argv(&[flag, "10.0.0.4:27015"])),
            Some(Invite::Address("10.0.0.4:27015".to_string())),
            "{flag}"
        );
    }
}

#[test]
fn a_truncated_command_line_produces_no_join_at_all() {
    // A flag with nothing after it must not become a join attempt against "" —
    // which is a probe the player never asked for and a refusal they cannot
    // read.
    assert_eq!(read_invite(&argv(&["+connect_lobby"])), None);
    assert_eq!(read_invite(&argv(&["--connect", "--fullscreen"])), None);
    assert_eq!(
        read_invite(&argv(&["+connect_lobby", "+connect_lobby"])),
        None
    );
}

#[test]
fn an_ordinary_launch_asks_for_nothing() {
    assert_eq!(read_invite(&argv(&[])), None);
    assert_eq!(read_invite(&argv(&["--multiplayer", "--mods"])), None);
}

#[test]
fn the_invite_is_found_wherever_the_platform_put_it() {
    // Steam appends its own arguments after the game's, and a launcher may put
    // them anywhere; the flag is searched for rather than expected first.
    assert_eq!(
        read_invite(&argv(&[
            "--multiplayer",
            "+connect_lobby",
            "42",
            "--verbose"
        ])),
        Some(Invite::Lobby("42".to_string()))
    );
}
